Fix issue DAN-169

akashniral opened the pull request:
1. Added link parameters to Danos yang model.

Signed-off-by: akashniral <akash@niralnetworks.com>

akashniral committed:
Fix issue DAN-169

1. Added link parameters to Danos yang model.

Signed-off-by: akashniral <akash@niralnetworks.com>

wivory commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+			}
+			leaf enable {
+				type empty;
+				must "(../enable)" {
```
leaf must be configured if it is configured - well yes, that's a racing certainty, but doesn't really add a lot to the model.  I'm not clear why you need the enable leaf at all to be honest - why would you need to enable *params*? If any are set, then they are set; if not, they're not set. Is this simply so your code can check the 'enable' leaf to determine whether to look for other params? Or am I missing something here?

akashniral replied:
In the updated code the must statement has been removed.

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+
+  revision 2021-03-02 {
+    description 
+      "Added link-params option.
```
Has this been tested at all? This description is never closed, so the file doesn't compile.

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+      		pattern "0[Xx][a-fA-F0-9]+";
+    		}
+				must "(../enable)" {
+					error-message "link-params enable must be set."
```
All of these are missing semicolons.

akashniral replied:
The missing semicolons have been added here.

JammyStuff reviewed (changes requested):
As-is, the YANG does not compile. Please fix the syntax errors and I'll review again.

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+  augment /if:interfaces/interfaces-dataplane:dataplane {
+    uses link-params;
+  }
+  augment /if:interfaces/interfaces-dataplane:dataplane/interfaces-dataplane:vif/interfaces-dataplane {
```
This path is invalid

akashniral replied:
The path is now correct after adding the :ip at the end of the path

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+module vyatta-protocols-frr-zebra-v1 {
```
Generally the names used in the YANG should not be tied to the implementation. Call this something other than zebra, maybe something like link-params.

akashniral replied:
I have changed the file name to vyatta-protocols-frr-link-params-v1.yang

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+  namespace "urn:vyatta.com:mgmt:vyatta-protocols-frr-zebra:1";
+  prefix vyatta-protocols-frr-zebra-v1;
+
+  import vyatta-protocols-v1 {
```
This doesn't seem to be used.

akashniral replied:
I have removed the unused imports i.e vyatta-protocols-v1

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+      "Added link-params option.";
+  }
+  
+	typedef ieee-float {
```
Both this and the float type below seem needlessly complicated. What's the intention here? Why was this method used rather than using a decimal64? Why are prefixing with 0x in some cases and not in others? The data model should abstract complexities in the underlying implementation, not expose them

akashniral replied:
The prefix 0x was used so that we even can take hex decimal format as inputs. In FRR the input format takes both hex and floating-point and is of type IEEE-float. Should I stick to decimal64 instead of the above float types?
![image](https://user-images.githubusercontent.com/79829103/117123915-46983c80-adb5-11eb-93f1-4c527c02e802.png)

akashniral replied:
Eg:
1.
![image](https://user-images.githubusercontent.com/79829103/117147909-54f55100-add3-11eb-8a42-9246492d57f6.png)
2.
![image](https://user-images.githubusercontent.com/79829103/117148334-c2a17d00-add3-11eb-8611-5255c5c6bdd0.png)
In Frr we were using both hex format and decimal format for taking input values like here in available bandwidth.
Should I just stick to floating-point instead of taking hex format for Danos CLI? What are your thoughts on this?

JammyStuff replied:
The only difference between your ieee-float and float types is the 0x in front, both of them seem to accept hex digits. 
 What is the difference between those types supposed to be? Configuring a bandwidth in hexadecimal also seems very strange, do you know of anyone that actually does that?

Thinking about it more, I'm not even convinced that a decimal type should be required here. If there's a suitable unit that could be used in all cases, you could just have a uint with that unit. If not, possibly something like mutually exclusive containers named after the unit containing a leaf with the value would be a good way of modelling this. @wivory, @jsouthworth any thoughts?

akashniral replied:
People do not generally use hexadecimal for setting up the values. Earlier I tried to keep the existing functionality as per FRR. However, I feel a simple uint32 would do and fit in all cases but I am open to hearing if you have any other suggestions @wivory, @jsouthworth.

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+module vyatta-protocols-frr-zebra-v1 {
```
You have a mixture of tabs and spaces in there. Please stick to tabs, and keep lines to 80 characters wide when the tabs are represented as 4 spaces wide.

akashniral replied:
The file has been converted to tabs

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+  grouping link-params {
+    container link-params {
+      configd:help "Configure interface link parameters";
+			leaf admin-grp {
```
Go with admin-group, shortening to grp is unclear and unnecessary.

akashniral replied:
renamed to admin-group

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+      	configd:help "Administrative group membership";
+				description "Administrative group membership";
+			}
+			leaf enable {
```
Why is there an explicit enable here? What does this add?

akashniral replied:
The explicit enable has been given as per the functionality stated in FRR i.e for activating link parameters on the interface

JammyStuff replied:
The model should model the concept well, not the underlying implementation. Are users likely to want to leave the parameters configured without them taking effect? My guess would be probably not, but I'm opening to hear if you have other options on that.

My preference would be dropping this entirely, but if there is use in users being able to disable the parameters without unconfiguring them, this should be done in a similar way to firewall rules where there is a disable rather than an enable (or an enum of enabled/disabled with a default of enabled).

akashniral replied:
Yes, I agree. I have dropped the explicit enable in the latest commit.

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+				description "Activate link parameters on this interface";
+			}
+
+			leaf ava-bw {
```
available-bandwidth

akashniral replied:
renamed to available-bandwidth

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+      	configd:help "Unidirectional Link Delay Variation";
+				description "Unidirectional Link Delay Variation";
+			}
+			leaf max-bw {
```
max-bandwidth

akashniral replied:
renamed to max-bandwidth

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+      	configd:help "Maximum bandwidth that can be used";
+				description "Maximum bandwidth that can be used";
+			}
+			leaf max-rsv-bw {
```
max-reserved-bandwidth

akashniral replied:
renamed to max-reserved-bandwidth

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+				description "Configure remote ASBR information (Neighbor IP address and AS number)";
+
+			}
+			leaf res-bw {
```
residual-bandwidth

akashniral replied:
renamed to residual-bandwidth

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+      	configd:help "Unidirectional Residual Bandwidth";
+				description "Unidirectional Residual Bandwidth";
+			}
+			leaf unrsv-bw {
```
unreserved-bandwidth

akashniral replied:
renamed to unreserved-bandwidth

JammyStuff commented on yang/vyatta-protocols-frr-zebra-v1.yang:
```diff
@@ -0,0 +1,227 @@
+      	configd:help "Unreserved bandwidth at each priority level";
+				description "Unreserved bandwidth at each priority level";
+			}
+			leaf use-bw {
```
utilized-bandwidth

akashniral replied:
renamed to utilised-bandwidth

JammyStuff reviewed (changes requested):
You've committed a bunch of unrelated stuff here, please clear it up.

akashniral committed:
Fix issue DAN-169

1. remove enable from danos cli
2. change the type from ieee-float to uint32

Signed-off-by: akashniral <akash@niralnetworks.com>

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,194 @@
+
+	revision 2021-03-02 {
+		description 
+			"Added link-params option.";
```
'Initial revision' would seem more appropriate.

akashniral replied:
changed to Initial revision

akashniral replied:
The description has been modified to "Initial revision"

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,194 @@
+				type string {
+					pattern "0[Xx][a-fA-F0-9]+";
+				}
+				must "(../enable)" {
```
Does this not require an 'enable' leaf in this container, which is missing? Did you mean '../../enable' or would you be better removing all these musts, and having it so any link-param configuration here implicitly enables link-params? Or make it a presence container so configuration of the container with no child nodes means link-params are enabled?

akashniral replied:
The must statement has been removed in the updated code.

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,194 @@
+			}
+			leaf metric {
+				type uint32 {
+					range "1..4294967295";
```
I think this is equivalent '1..max' which might read better?

akashniral replied:
max would be better to read. It has been updated.

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,194 @@
+				must "(../enable)" {
+					error-message "link-params enable must be set.";
+				}
+				configd:help "Unidirectional Available Bandwidth";
```
You might want to give the user a clue about units? bits? Gigabytes?

akashniral replied:
The units have already been added i.e (Bytes/second).

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,194 @@
+			}
+			leaf delay {
+				type uint32 {
+					range "0..16777215";
```
nanoseconds? Years?

akashniral replied:
The units have already been added in the previous commit i.e (micro-second).

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,194 @@
+			}
+			leaf delay-variation {
+				type uint32 {
+					range "0..16777215";
```
Units?

akashniral replied:
The units have already been added in the previous commit i.e (micro-second).

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,194 @@
+				description "Unidirectional Link Delay Variation";
+			}
+			leaf max-bandwidth {
+				type uint32;
```
Units

akashniral replied:
The units have already been added in the previous commit i.e (Bytes/second).

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,194 @@
+				description "Maximum bandwidth that can be used";
+			}
+			leaf max-reserved-bandwidth {
+				type uint32;
```
Units

akashniral replied:
The units have already been added in the previous commit i.e (Bytes/second).

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,194 @@
+				description "Configure remote ASBR information (Neighbor IP address and AS number)";
+			}
+			leaf packet-loss {
+				type uint32;
```
Units

akashniral replied:
The units have already been added in the previous commit i.e (Percentage).

JammyStuff commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,194 @@
+				type string {
+					pattern "0[Xx][a-fA-F0-9]+";
+				}
+				must "(../enable)" {
```
Given that enable is gone, these must statements don't seem right (on all of the leaf nodes below as well).

akashniral replied:
I've removed the must statement as enable was gone

JammyStuff commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,194 @@
+				description "Administrative group membership";
+			}
+			leaf available-bandwidth {
+				type uint32;
```
Much cleaner. Please add units statements to all of these so it's clear what units are being used. Maybe also mention the units in the configd:help as units aren't exposed via the CLI help.

akashniral replied:
added units statements in each of the leaf nodes and configd:help

wivory replied:
So you're never going to exceed ~4GB? Would a uint64 be safer here? Same applies to all bandwidth fields

akashniral replied:
uint64 would be able to allocate more bandwidth than uint32.. It has been modified

wivory replied:
Please push changes - presumably you will be modifying ALL bandwidth nodes to uint64?

akashniral replied:
I've pushed my changes. In all the bandwidth I've modified to uint64

JammyStuff reviewed (changes requested):
Looking much better. Once these are fixed, unless anything else comes up that needs fixing it would be good to see some testing showing that this works, as there's been a few iterations presented for review that wouldn't have worked.

Also there's now a conflict in the control file, which will need resolving.

akashniral committed:
Fix issue DAN-169

1. remove must statement
2. add units statement in each of the leaf nodes and in confid:help

Signed-off-by: akashniral <akash@niralnetworks.com>

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,170 @@
+		Web: www.niralnetworks.com";
+
+	description
+		"Copyright (c) 2019-2020, Niral Networks Intellectual Property.
```
Should this just be 2021 if a new file? Certainly should list 2021

akashniral replied:
2021 has been listed i.e 2019-2021

akashniral commented:
> Looking much better. Once these are fixed, unless anything else comes up that needs fixing it would be good to see some testing showing that this works, as there's been a few iterations presented for review that wouldn't have worked.
> 
> Also there's now a conflict in the control file, which will need resolving.

akashniral committed:
Merge branch 'master' into niral_zebra_cli

akashniral commented:
> Looking much better. Once these are fixed, unless anything else comes up that needs fixing it would be good to see some testing showing that this works, as there's been a few iterations presented for review that wouldn't have worked.
> 
> Also there's now a conflict in the control file, which will need resolving.

The conflict is now resolved.

The following commands are now being shown in the Danos CLI for link-params with units
![image](https://user-images.githubusercontent.com/79829103/118263084-6fb27e80-b4d3-11eb-8d04-1120eb63e93f.png)

Below is the simple configuration for link-params
![image](https://user-images.githubusercontent.com/79829103/118263351-c5872680-b4d3-11eb-9ed5-31022f874872.png)

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,170 @@
+			}
+			leaf unreserved-bandwidth {
+				type uint32 {
+					range "0..7";
```
Units?

akashniral replied:
Here the priority level is been set i.e 0-7. I've added the units as "Priority"

wivory replied:
(a) you haven't pushed your changes

(b) even if you did, this still makes no sense. How can a single priority level (misnamed as 'unreserved-bandwidth') represent 'unreserved bandwidth at each priority level'? Are you trying to set an amount of bandwidth for each of the 8 priority levels? You can't do that with a single value ...

akashniral replied:
I have changed the yang to set the amount of bandwidth for each of the 8 priority levels.
The command is as follows: set interface dataplane dp0xxx link-params unreserved-bandwidth priority <x> bandwidth <y>. However, in the frr the command is direct for setting the bandwidth i.e: unrsv-bw <priority level 0-7> <bandwidth>.

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,170 @@
+				configd:help "Maximum bandwidth that can be used (Bytes/second)";
+				description "Maximum bandwidth that can be used";
+			}
+			leaf max-reserved-bandwidth {
```
Can this exceed max-bandwidth? If not then perhaps a must statement enforcing this would be wise.

wivory replied:
How do these 2 relate to available-bandwidth, residual, unreserved and utilised?

akashniral replied:
1. I've enforced must statement in each of the bandwidths. Before configuring other bandwidths the max-bandwidth has to be set in order to compare the values i.e (max-bandwidth >= <other-related-bandwidths-fields>).
2. I have described the bandwidth-related fields following draft-ietf-isis-te-metrics-extension-03.

wivory replied:
What determines the max bandwidth that can be used? Is this an inherent property of the link, or, like ADSL, depends on how well the link is working at a given time?

akashniral replied:
I am not sure about it. In frr the max-bandwidth is set upon setting link-param however in Danos CLI user can provide max-bandwidth according to the need.

wivory replied:
How can you code a feature when you don't know what it means???

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,170 @@
+				configd:help "Maximum bandwidth that may be reserved (Bytes/second)";
+				description "Maximum bandwidth that may be reserved";
+			}
+			leaf metric {
```
Could you ever have other metrics? Maybe mpls-te-metric would be a better name?

akashniral replied:
It seems better.. It has been changed

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,170 @@
+					range 0..100;
+				}
+				units "Percentage";
+				configd:help "Configure remote ASBR information (Neighbor IP address and AS number) in percentage";
```
Cut and paste from previous element? Just adding 'in percentage' does not make it make sense.

akashniral replied:
The configd:help and description has been changed to Unidirectional Link Packet Loss (Percentage).

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,170 @@
+				type uint32;
+				units "Bytes/second";
+				configd:help "Unidirectional Residual Bandwidth (Bytes/second)";
+				description "Unidirectional Residual Bandwidth";
```
What is residual bandwidth? Description can (and often should) be multi-line to give a fuller description than configd:help.

akashniral replied:
Fuller description has been provided for residual bandwidth following draft-ietf-isis-te-metrics-extension-03.

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,170 @@
+				configd:help "Unreserved bandwidth at each priority level";
+				description "Unreserved bandwidth at each priority level";
+			}
+			leaf utilised-bandwidth {
```
Is this config or state? Please provide a better description.

akashniral replied:
Fuller description has been provided for utilised bandwidth following draft-ietf-isis-te-metrics-extension-03.

wivory replied:
This still sounds more like state than config - 'as measured in the router' is state not config AFAICT.

akashniral replied:
This similar command is there in vtysh shell where I can use this to adjust utilized-bandwidth in Bytes/second and was req in Danos CLI and in the draft-ietf-isis-te-metrics-extension-03 the above description is given of the same therefore I've provided that description.

akashniral committed:
Fix issue DAN-169

modified the uint32 to uint64 for allocating more bandwidth greater than 4GB

Signed-off-by: akashniral <akash@niralnetworks.com>

akashniral committed:
Merge branch 'niral_zebra_cli' of https://github.com/Niral-Networks/vyatta-protocols-frr into niral_zebra_cli

akashniral committed:
Fix issue DAN-169

1. add description in each of the bandwidth releated fields.
2. enforce must statement so other bandwidth related field do not exceed maximum bandwidth
3. change unreserved bandwidth field to set the amount of bandwidth for each of the priority levels.

Signed-off-by: akashniral <akash@niralnetworks.com>

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,229 @@
+				configd:help "Unidirectional Residual Bandwidth (Bytes/second)";
+				description "Unidirectional Residual Bandwidth. For a link or forwarding adjacency, residual
+					bandwidth is defined to be Maximum Bandwidth [RFC3630] minus the bandwidth currently allocated
+					to RSVP-TE LSPs. For a bundled link, residual bandwidth is defined to be the sum of the
```
Where is the bandwidth allocated to RSVP-TE LSPs defined? What if there is a mismatch?

akashniral replied:
I have found this description in draft-ietf-isis-te-metrics-extension-03. Do I have to define bandwidth allocated to RSVP-TE LSP too in this DAN-169 fix?

wivory replied:
You aren't convincing me that you fully understand what all these fields mean, nor how they relate to each other. I am not convinced that this data model is complete / consistent, and until I am persuaded that it is, I cannot approve.

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,229 @@
+					error-message "Residual Bandwidth cannot be greater than Maximum Bandwidth";
+				}
+			}
+			container unreserved-bandwidth {
```
If max-bandwidth is X, it seems you can have X unreserved bandwidth at each of the 8 priority levels, making 8X?

akashniral replied:
![image](https://user-images.githubusercontent.com/79829103/119960019-29284e00-bfc2-11eb-8235-52f72ab8fd57.png)
In each of the priority levels, the bandwidth can be set equal to max-bandwidth. The same way has been implemented in FRR.

wivory replied:
That's as maybe. I'd like you to explain to me why this is correct, rather than just saying it's done this way somewhere else.

akashniral replied:
The [RFC3630] has been followed here wherein section 2.5.8 it stated about Unreserved Bandwidth as follows:

The Unreserved Bandwidth sub-TLV specifies the amount of bandwidth not yet reserved at each of the eight priority levels in IEEE floating-point format.  The values correspond to the bandwidth that can be reserved with a setup priority of 0 through 7, arranged in increasing order with priority 0 occurring at the start of the sub-TLV, and priority 7 at the end of the sub-TLV.  The initial values (before any bandwidth is reserved) are all set to the Maximum Reservable Bandwidth.  Each value will be less than or equal to the Maximum Reservable Bandwidth.  The units are bytes per second.
The Unreserved Bandwidth sub-TLV is TLV type 8, and is 32 octets in length.

wivory replied:
I'm still not convinced this is config as opposed to state information.

wivory replied:
or at least information that should be dynamically derived not set in stone in the config

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,229 @@
+			container unreserved-bandwidth {
+                                configd:help "Unreserved bandwidth at each priority level";
+                                description "Unreserved bandwidth at each priority level. Unreserved Bandwidth [RFC3630], is
+					subtracted from the Maximum Reservable Bandwidth (the bandwidth that can theoretically
```
bad alignment

akashniral replied:
I've aligned the description.

wivory commented on yang/vyatta-protocols-frr-link-params-v1.yang:
```diff
@@ -0,0 +1,229 @@
+                                                units "Bytes/second";
+                                                configd:help "Bandwidth (Bytes/second)";
+                                                description "Bandwidth (Bytes/second) for each priority levels";
+						must "../../../max-bandwidth" {
```
bad alignment

akashniral replied:
It has been aligned

wivory replied:
Still looks to be 2 spaces too far to the left ...

akashniral replied:
The changes have been reflected in the last commit

akashniral committed:
Fix issue DAN-169

align description field

Signed-off-by: akashniral <akash@niralnetworks.com>

akashniral committed:
Fix issue DAN-169

fix alignment issues

Signed-off-by: akashniral <akash@niralnetworks.com>

wivory reviewed (commented):
I don't feel that the submitter understands the interaction between all the different bandwidth fields. Simply cutting and pasting descriptions from another document is not enough. I cannot approve until the submitter has satisfactorily explained to me how they all fit together.
